### PR TITLE
Interrupts on RISC-V

### PR DESCRIPTION
**Pull request by thejpster**
) Adds an Xh3irq controller driver
) Adds a default `MachineExternal` handler which uses the Xh3irq to run whichever interrupts are pending
) Adds a bunch of interrupt control stuff to `hal::arch`
) Fixes the `powman_test` example to use the new interrupt APIs - it now works on RISC-V and Arm.

Tested `powman_test` on a Pico 2 in both `riscv32imac-unknown-none-elf` and `thumbv8m.main-none-eabihf`.

### CODE REVIEW

File: rp235x-hal-examples/src/bin/i2c_async.rs [hunk 1]
```diff
@@ -98,8 +98,8 @@ async fn demo() {
     // Each core has its own NVIC so these needs to executed from the core where the IRQ are
```
Comment (9names) [inline rp235x-hal-examples/src/bin/i2c_async.rs:98]: Not really an NVIC here, is it?

File: rp235x-hal/src/xh3irq.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        assert_eq!(index, 2);
+        assert_eq!(bitmask, 0b0000_0000_0000_0100);
+    }
+
+    #[test]
+    fn test_interrupt_to_mask_index() {
+        let mask = interrupt_to_mask_index(rp235x_pac::Interrupt::UART1_IRQ);
+        // 34 -> bit 3 | window 2
+        assert_eq!(mask, 0x0004_0002);
+    }
```
Comment (9names) [inline rp235x-hal/src/xh3irq.rs:236]: what does the 34 comment mean here?

Comment (thejpster) [reply rp235x-hal/src/xh3irq.rs:236]: Oh, UART1_IRQ has the value 34

**Comment by thejpster** [comment]
We might want to block this until we've written a new interrupt macro that works like the cortex-m-rt macro and does the static mut transform and type-checks the function signature.

File: rp235x-hal-examples/src/bin/i2c_async.rs (hunk 1 above)
Comment (thejpster) [reply rp235x-hal-examples/src/bin/i2c_async.rs]: Cleaned up the comment

File: rp235x-hal/src/xh3irq.rs (hunk 2 above)
Comment (thejpster) [reply rp235x-hal/src/xh3irq.rs]: Cleaned up the test and used another IRQ where the bit number and the index have different numeric values

File: rp235x-hal-examples/Cargo.toml [hunk 3]
```diff
@@ -28,10 +28,15 @@ futures = {version = "0.3.30", default-features = false, features = ["async-awai
...
 pio-proc = "0.2.0"
 rp235x-hal = {path = "../rp235x-hal", version = "0.2.0", features = ["binary-info", "critical-section-impl", "rt", "defmt"]}
 usb-device = "0.3.2"
 usbd-serial = "0.2.2"
+
+[target.'thumbv8m.main-none-eabihf'.dependencies]
+nostd_async = {version = "0.6.1", features = ["cortex_m"]}
+
+[target.riscv32imac-unknown-none-elf.dependencies]
+nostd_async = {version = "0.6.1"}
```
Comment (jannic) [inline rp235x-hal-examples/Cargo.toml:42]: I don't think `nostd_async` should be used without the `cortex_m` feature. The fallback implementation of `interrupt::free` just doesn't disable any interrupts and unsafely creates a `CriticalSection` token anyway:

https://github.com/sammhicks/nostd_async/blob/master/src/interrupt.rs#L5-L10

Comment (thejpster) [reply rp235x-hal-examples/Cargo.toml:42]: Yeah. Not sure what we should do instead though.

Comment (jannic) [reply rp235x-hal-examples/Cargo.toml:42]: I'd say either embassy-executor (AFAIKT the most popular no_std executor), or some minimal toy executor like the one I suggested for on-target tests in #848.

There are two conflicting goals when choosing the executor to be used for the examples:
- Examples should be concise
- Examples should introduce knowledge that's applicable to real-world tasks

Depending on what you value more, you could justify both choices.
In any case, nostd_async IMHO isn't a good choice, as it may look like it's usable for real (non-example) code, while in fact it has severe issues.

I'm not knowledgeable enough in writing async executors to decide if a third option could be viable: Improving nostd_async. It may only need a few small fixes, or it may be fundamentally flawed - I just don't know.

File: rp235x-hal/src/arch.rs [hunk 4]
```diff
@@ -73,43 +101,476 @@ mod inner {
...
+    /// Just panics.
+    ///
+    /// # Safety
+    ///
+    /// Do not call this function - it is called automatically when our
+    /// `MachineExternal` interrupt handler can't find anything better to call.
+    #[no_mangle]
+    #[allow(non_snake_case)]
+    unsafe fn DefaultIrqHandler() {
+        panic!();
```
Comment (jannic) [inline rp235x-hal/src/arch.rs:485]: cortex_m does not panic but `loop {}` in the default handler.
It did panic for a short while in the past, but that was reverted here: https://github.com/rust-embedded/cortex-m-rt/pull/289/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759L993

I'm not sure if the reasons for that decision also apply for RISC-V, but at least for the rp2040/rp235x case it would be nicer if both implementations did the same.

**Comment by jannic** [comment]
> We might want to block this until we've written a new interrupt macro that works like the cortex-m-rt macro and does the static mut transform and type-checks the function signature.

As the rp235x is a multi-core processor, and given that https://github.com/rust-embedded/cortex-m/issues/411 indicates that the transformation made by cortex-m-rt is unsound on multi-core systems, I don't think we should just copy it. Is there a good way to make the transform safe on rp235x? Taking a full critical section on each interrupt entry is probably to expensive?

File: rp235x-hal-examples/src/bin/gpio_irq_example.rs [hunk 5]
```diff
@@ -142,21 +144,23 @@ fn main() -> ! {
     }
 }
 
-#[interrupt]
+#[no_mangle]
+#[allow(non_snake_case)]
 fn IO_IRQ_BANK0() {
     // The `#[interrupt]` attribute covertly converts this to `&'static mut Option<LedAndButton>`
     static mut LED_AND_BUTTON: Option<LedAndButton> = None;
+    let led_and_button = unsafe { &mut *core::ptr::addr_of_mut!(LED_AND_BUTTON) };
```
Comment (jannic) [inline rp235x-hal-examples/src/bin/gpio_irq_example.rs:152]: This `&mut *core::ptr::addr_of_mut!( ... )` IMHO doesn't make sense.

The whole raison d'être of `addr_of_mut!` is avoiding the `&mut`, because having a `&mut` (even for a short time, and not derefencing it) can potentially cause UB if you are not careful.

But here, you convert it to a `&mut` anyway. So I don't see any advantage of this difficult to read incantation, you can as well just do `unsafe { &mut LED_AND_BUTTON };` instead.

Now, that causes a warning. And this warning looks like it led you to writing the code above:

```
warning: creating a mutable reference to mutable static is discouraged
   --> src/bin/gpio_irq_example.rs:152:35
    |
152 |     let led_and_button = unsafe { &mut LED_AND_BUTTON };
    |                                   ^^^^^^^^^^^^^^^^^^^ mutable reference to mutable static
    |
    = note: for more information, see issue #114447 <https://github.com/rust-lang/rust/issues/114447>
    = note: this will be a hard error in the 2024 edition
    = note: this mutable reference has lifetime `'static`, but if the static gets accessed (read or written) by any other means, or any other reference is created, then any further use of this mutable reference is Undefined Behavior
    = note: `#[warn(static_mut_refs)]` on by default
help: use `addr_of_mut!` instead to create a raw pointer
    |
152 |     let led_and_button = unsafe { addr_of_mut!(LED_AND_BUTTON) };
    |                                   ~~~~~~~~~~~~~              +
```

I think the warning is valid, but the `help` is misleading. It even says "to create a raw pointer", and as written it would return a pointer, not a `&mut`, so the resulting code wouldn't build.

TL;DR:

```suggestion
    #[allow(static_mut_refs)]
    // Safety: As long as this interrupt is only enabled on a single core,
    // and the interrupt isn't somehow called reentrantly, there will not
    // be a second reference to LED_AND_BUTTON at the same time.
    let led_and_button = unsafe { &mut LED_AND_BUTTON };
```

**Comment by thejpster** [comment]
So I guess the risk is that both CPUs enter the same interrupt handler at the same time. If one peripheral interrupt signal is unmasked on both CPUs, that could happen. So perhaps the fix is to design the interrupt API to make that impossible (like a bitmask in a static atomic or the watchdog scratch that records which cpu has which interrupt enabled). Or we could ensure that if you enter an interrupt on Core A, it is masked for the duration on Core B - which is nicer than turning all interrupts off or holding a spin lock.

But given 2040 has the same problem the question is whether we want to fix it here or make a plan and fix it later.

**Comment by 9names** [comment]
The user already unmasks the interrupt, and that operation is unsafe. It is almost certainly a bug if they do so, and we have had no reports of people doing this accidentally.

Until someone comes has a use-case where having both cores handle an interrupt is beneficial, I think trying to solve this is not a good use of our time.

My opinion is we should document "don't do this" and move on with our lives

**Comment by jannic** [comment]
> But given 2040 has the same problem the question is whether we want to fix it here or make a plan and fix it later.

That's an option, I just wouldn't want to spend time on porting the macro only to remove it again later.

> Until someone comes has a use-case where having both cores handle an interrupt is beneficial,

Are there situations where it's not avoidable? Any non maskable interrupts or exceptions?

**Comment by jannic** [comment]
But to be honest, there's a second thing I dislike about the interrupt macro, so I'm biased.

For me, it's totally non-obvious that the macro annotation changes the code inside the function.
So if I see this:
```
#[interrupt]
fn SOME_INTERRUPT() {
  static mut SOMETHING: Option<Whatever> = None;
  [...]
  *SOMETHING = foo(); 
}
```
I always feel like the usage of the variable doesn't match the declaration. This was very confusing when I was new to embedded rust, and it still sometimes surprises me.

Therefore I'd prefer something more obvious. Perhaps a macro directly on the static mut?
```
#[interrupt]
fn SOME_INTERRUPT() {
  let something = interrupt_singleton!( didn't think about what comes here yet :Option<Whatever> );
  [...]
  *something = foo(); 
}
```

Probably needs an `unsafe` declaration as well, as using this macro outside an interrupt (or in a re-entrant interrupt) would be unsound. So it's far from a finished proposal, of course. This is only meant as an explanation why I don't like the current interrupt macro.

Back to topic: Personally I'd merge the PR as it is, and not wait for an `#[interrupt]` macro implementation for RISC-V. The current situation is not perfect, but it's good enough, and it's not clear yet how an improvement would look like.

(That said, if someone ported `#[interrupt]` I wouldn't oppose that either. While I still don't like it, I do value consistency.)

**Comment by jannic** [comment]
> So I guess the risk is that both CPUs enter the same interrupt handler at the same time. If one peripheral interrupt signal is unmasked on both CPUs, that could happen.

BTW, do I understand `xh3irq` correctly that it already solves this issue? `xh3irq::get_next_interrupt()` should only return a given interrupt to a single CPU core, even if the interrupt is enabled on both, right?

File: rp235x-hal-examples/Cargo.toml (hunk 3 above)
Comment (jannic) [reply rp235x-hal-examples/Cargo.toml:42]: https://github.com/rp-rs/rp-hal/pull/850

**Comment by thejpster** [comment]
> > So I guess the risk is that both CPUs enter the same interrupt handler at the same time. If one peripheral interrupt signal is unmasked on both CPUs, that could happen.
> 
> 
> 
> BTW, do I understand `xh3irq` correctly that it already solves this issue? `xh3irq::get_next_interrupt()` should only return a given interrupt to a single CPU core, even if the interrupt is enabled on both, right?

I think the UART interrupt is cleared by reading the FIFO which can happen before the ISR ends. So it could re-fire before the ISR ends. Not sure either Interrupt Controller can avoid that.

**Comment by thejpster** [comment]
What do we want to do with this? There's been some changes over in riscv-rt since.

**Comment by winksaville** [comment]
> What do we want to do with this? There's been some changes over in riscv-rt since.

I'd like to see interrupts working on risc-v and willing to work on this if you'd like. That
said, I'd need some initial guidance as I'm a noob so it will definitely take me some time.

File: rp235x-hal/src/arch.rs (hunk 4 above)
Comment (thejpster) [reply rp235x-hal/src/arch.rs:485]: I'll fix this

File: rp235x-hal-examples/src/bin/gpio_irq_example.rs (hunk 5 above)
Comment (thejpster) [reply rp235x-hal-examples/src/bin/gpio_irq_example.rs:152]: This will be a hard error in 2024 Edition, so I'll think of a better solution.

**Comment by jonathanpallant** [comment]
I rebased and took out any use of the unsound static-mut conversion. Now the interrupts just run in a critical section. Not yet tested on hardware.

File: rp235x-hal/src/arch.rs (hunk 4 above)
Comment (jonathanpallant) [reply rp235x-hal/src/arch.rs]: Changed it to spin

**Comment by thejpster** [comment]
What do we need in order to merge this?

**Comment by 9names** [comment]
> Not yet tested on hardware.

Might just be me, but some proof (even anecdotal) that it works on the hardware it's supposed to target would be nice before merging it.

**Review by 9names — APPROVED** [review]
I tested that both the gpio_irq_example and powman_test examples worked on both riscv32imac-unknown-none-elf and thumbv8m.main-none-eabihf on a Pico 2.

LGTM

**Comment by thejpster** [comment]
Thank you for checking on hardware. Much appreciated. My Pico 2 is under a pile of other stuff at the moment, sadly.
